Add support for multiplatform builds

alfrunes opened the pull request:
See `README.make`.

alfrunes committed:
build(make): Introduce better make variables for docker builds

This commit changes the make variable interface for building docker
images.
 backend/Makefile:
  * MENDER_IMAGE_PREFIX sets the docker registry and path prefix for
    docker builds (default: localhost:5000/mender-server)
  * MENDER_IMAGE_TAG defines the tag name (default: latest)
backend/services/Makefile.common:
  * DOCKER_PLATFORM is passed directly to docker build `--platform`
    argument. The default value is inferred from `docker system info` to
    avoid building on wrong target OS when building on Docker Desktop
    for Mac.
backend/services/*/Makefile:
  * build-test target has been removed
  * docker-acceptance target has been updated to build with docker
    target with a fixed set of build variables.
  * DOCKER_ARGS changed name to DOCKER_BUILD_ARGS

NOTE: Building for multiple platforms usually does not work using the
default docker builder. The build context must be updated for the docker
targets to work.

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
chore(dev): Align compose images with Makefile defaults

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
build(make): Iniitalize root Makefile and frontend/Makefile

The root makefile only exposes the top level targets `build` and
`docker` for now and calls the respective targets in the backend and
frontend.

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
build(make): Remove distdir and do not create subdirectories for os/arch

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
docs(make): Add README.make for documenting the build targets

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
ci: Add job for building docker images from binaries

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
ci: Remove go dependency from running acceptance tests

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

mzedel commented on .gitlab-ci.yml:
```diff
@@ -44,6 +48,22 @@ build:backend:dist:
+  tags:
+    - hetzner-amd-beefy
+  script:
+    - make docker
```
do we really want to rebuild all the services for a single service change? This will unnecessarily slow down the pipeline

oldgiova replied:
Additionally, it could be wonderful if we can use parallel jobs when we have to build multiple changed services.

mzedel commented on frontend/Makefile:
```diff
@@ -0,0 +1,26 @@
+
+build:
+	npm install && \
+    npm run build
```
```suggestion
build:
	npm ci && \
    npm run build
```
otherwise you won't get reliably reproducible results.

mzedel commented on .gitlab-ci.yml:
```diff
@@ -44,6 +48,22 @@ build:backend:dist:
     paths:
       - dist.tar.gz
 
+build:docker:
```
```suggestion
build:backend:docker:
```

alfrunes committed:
build(make): Remove debug prints

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
chore: Clean up unused acceptance tests compose files for enterprise

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
build(make): Add build dependencies and update `%-acceptance` targets

The `build` targets tracks the target binary and depends on the source
files. This prevents make from attempting to rebuild the binary when
building the docker image if the binary is already up to date.

The `docker-acceptance` target is turned into a simple dependency on
docker with build flags for adding coverage and other defaults for
docker tag.

The top level `test-acceptance` no longer tears down the compose
environment, but leaves the existing containers running for faster
execution of acceptance tests.

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
ci: Do not propagate build artifacts (binaries) to test stage

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
ci: Exclude frontend and rename `build:backend:*` jobs

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
ci: Use correct image when building docker images

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
ci: Bump docker major version to 27

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
build(make): Do not infer compoenent name using `go list`

Makes all targets dependent on `go`.

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
build(make): Target `build` makes all target platforms

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
build(make): Infer target arch from DOCKER_PLATFORM on `build`

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes committed:
ci: Initialize multiplatform builder before building containers

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>
